src/other/tabletraits.jl

```diff
@@ -0,0 +1,154 @@
+        if isa(df.columns[i], DataArray)
+            push!(col_expressions, Expr(:(::), names(df)[i], DataValues.DataValue{eltype(df.columns[i])}))
+        else
+            push!(col_expressions, Expr(:(::), names(df)[i], eltype(df.columns[i])))
+        end
+        push!(df_columns_tuple_type.args, typeof(df.columns[i]))
+    end
+    t_expr = NamedTuples.make_tuple(col_expressions)
+
+    t2 = :(DataFrameIterator{Float64,Float64})
+    t2.args[2] = t_expr
+    t2.args[3] = df_columns_tuple_type
+
+    t = eval(t2)
+
+    e_df = t(df, (df.columns...))
+
+    return e_df
+end
+
+function Base.length{T,TS}(iter::DataFrameIterator{T,TS})
+    return size(iter.df,1)
+end
+
+function Base.eltype{T,TS}(iter::DataFrameIterator{T,TS})
+    return T
+end
+
+function Base.start{T,TS}(iter::DataFrameIterator{T,TS})
+    return 1
+end
+
+@generated function Base.next{T,TS}(iter::DataFrameIterator{T,TS}, state)
+    constructor_call = Expr(:call, :($T))
+    for i in 1:length(iter.types[2].types)
+        if iter.parameters[1].parameters[i] <: DataValues.DataValue
+            push!(constructor_call.args, :(isna(columns[$i],i) ? $(iter.parameters[1].parameters[i])() : $(iter.parameters[1].parameters[i])(columns[$i][i])))
+        else
+            push!(constructor_call.args, :(columns[$i][i]))
+        end
+    end
+
+    quote
+        i = state
+        columns = iter.columns
+        a = $constructor_call
+        return a, state+1
+    end
+end
+
+function Base.done{T,TS}(iter::DataFrameIterator{T,TS}, state)
+    return state>size(iter.df,1)
+end
+
+# Sink
+
+@generated function _tabletraits_fill_df(columns, enumerable)
+    n = length(columns.types)
+    push_exprs = Expr(:block)
+    for i in 1:n
+        if columns.parameters[i] <: DataArray
+            ex = :( push!(columns[$i], isnull(i[$i]) ? DataArrays.NA : get(i[$i])) )
+        else
+            ex = :( push!(columns[$i], i[$i]) )
+        end
+        push!(push_exprs.args, ex)
+    end
+
+    quote
+        for i in enumerable
+            $push_exprs
+        end
+    end
+end
+
+function _construct_dataframe_from_iterabletable(x)
+    iter = TableTraits.getiterator(x)
+
+    T = eltype(iter)
+    if !(T<:NamedTuples.NamedTuple)
+        error("Can only collect a NamedTuple iterator into a DataFrame")
+    end
+
+    column_types = TableTraits.column_types(iter)
+    column_names = TableTraits.column_names(iter)
+
+    columns = []
+    for t in column_types
+        if isa(t, TypeVar)
+            push!(columns, Array{Any}(0))
+        elseif t <: DataValues.DataValue
+            push!(columns, DataArray(t.parameters[1],0))
+        else
+            push!(columns, Array{t}(0))
+        end
+    end
+    df = DataFrame(columns, fieldnames(T))
+    _tabletraits_fill_df((df.columns...), iter)
+    return df
+end
+
+function DataFrame(x::Array{T,1}) where {T<:NamedTuples.NamedTuple}
+    return _construct_dataframe_from_iterabletable(x)
+end
+
+function DataFrame(x)
```

Comment:
I feel like this is one of the most unfortunate parts of the IterableTables required interface; table types should be able to define their own catchall constructor without having to be required for IterableTables functionality. I think it'd be better to transition this to an `IterableTables.IterableTable(::Type{DataFrame}, x)` method that packages can implement.

The catchall constructor ends up getting called from [here](https://github.com/davidanthoff/Query.jl/blob/ca1b176326840c9e4c622fb3e068294fbdf88e27/src/sinks/sink_type.jl#L2), right? I think it'd be better to not have to rely on the catchall constructor there.

Reply:
This is not part of the required interface of IterableTables.jl, I only recommend it (I think that is pretty clear in the documentation). If sinks don't want to do this, they can do whatever else they want. The whole user API won't be as smooth for them, but they can still participate in the iterable tables story.

The question of how to integrate with Query.jl is to some extent distinct. In general you don't have to terminate a query with a ``@collect`` statement, and then you can just use whatever method you want to materialize things. But yes, the ``@collect T`` pattern requires that there is a constructor for ``T`` that can deal with whatever the query returns. The new piping syntax also only works if that is the case: ``df |> @where(_.a>2) |> DataFrame`` obviously requires a ``DataFrame`` constructor that can handle whatever is returned by ``@where``.

Reply:
C'mon, that's not really fair. Basically, if you don't, you can't use full Query functionality w/ your type, oh, but it's still optional? It really shouldn't be hard to change it to be a proper part of the interface, instead of requiring this "constructor coercion".

Reply:
Probably best to discuss this in an issue over in Query.jl, do you want to open an issue there?